patches/tModLoader/Terraria/ModLoader/NPCLoader.cs

```diff
@@ -401,19 +408,29 @@ public static void ReceiveExtraAI(NPC npc, byte[] extraAI)
 			}
 
 			if (stream.Position < stream.Length) {
-				throw new IOException($"Read underflow {stream.Length - stream.Position} of {stream.Length} bytes in ReceiveExtraAI, more info below");
+				// Calculate original length of Read7BitEncodedInt
+				int bitReaderOverhead = 1;
+				int temp = bitReader.MaxBits;
+				while (temp > 0x7Fu) {
+					bitReaderOverhead++;
+					temp >>= 7;
+				}
+				var bytesLength = stream.Length - (int)Math.Ceiling(bitReader.MaxBits / 8f) - bitReaderOverhead;
+				throw new IOException($"Read underflow {stream.Length - stream.Position} of {bytesLength} bytes in ReceiveExtraAI, more info below");
 			}
 		}
-		catch (Exception e) {
+		catch (Exception) {
 			string message = $"Error in ReceiveExtraAI for NPC {npc.ModNPC?.FullName ?? npc.TypeName}";
-			if (anyGlobals) {
+			if (lastGlobalNPC != null) {
 				message += ", may be caused by one of these GlobalNPCs:";
 				foreach (var g in HookReceiveExtraAI.Enumerate(npc)) {
 					message += $"\n\t{g.FullName}";
+					if (lastGlobalNPC == g)
+						break;
 				}
 			}
 
-			Logging.tML.Error(message, e);
```

Comment:
Please include the original exception. It may be a legitimate exception in a mod's netcode and we don't rethrow. (I made a mistake in my recommendation earlier, seeing it was unused in `ProjectileLoader`)

Reply:
The exception will already be logged before this message, so I figured we can omit it.

![image](https://github.com/user-attachments/assets/f0a3a54a-f494-48bf-befe-f2dc7af277a9)

Reply:
Hmm, I feel like relying on the `SilentlyCaughtException` is bad practice, but maybe I'm wrong